lib/chef/provider/aws_ebs_volume.rb

```diff
@@ -37,12 +42,115 @@ class Chef::Provider::AwsEbsVolume < Chef::Provisioning::AWSDriver::AWSProvider
     status = aws_object ? aws_object.status : nil
     case status
     when nil, :deleted, :deleting
+    when :in_use
+      current_attachment = aws_object.attachments.first
+      Chef::Log.info("EBS volume #{new_resource.name} (#{aws_object.id}) is attached to instance #{current_attachment.instance.id}. Detaching from instance #{current_attachment.instance.id}.")
+      detach(:instance => current_attachment.instance, :device => current_attachment.device)
+      delete
     else
-      converge_by "Deleting EBS volume #{new_resource.name} in #{region}" do
-        aws_object.delete
-      end
-      # TODO wait until deleted?
+      delete
     end
     new_resource.delete_managed_entry(action_handler)
   end
+
+  action :attach do
+    aws_object = new_resource.aws_object
+    status = aws_object ? aws_object.status : nil
+    case status
+    when :in_use
+      current_attachment = aws_object.attachments.first
+      instance = Chef::Resource::AwsInstance.get_aws_object(new_resource.machine, resource: new_resource)
+      # wrong instance
+      if current_attachment.instance != instance
+        Chef::Log.info("EBS volume #{new_resource.name} (#{aws_object.id}) is attached to instance #{current_attachment.instance.id}. Reattaching to instance #{instance.id} to device #{new_resource.device}.")
+        detach(:instance => current_attachment.instance, :device => current_attachment.device)
+        attach
+      # wrong device only
+      elsif current_attachment.instance == instance and current_attachment.device != new_resource.device
+        Chef::Log.info("EBS volume #{new_resource.name} (#{aws_object.id}) is attached to instance #{current_attachment.instance.id} on device #{current_attachment.device}. Reattaching device to #{new_resource.device}.")
+        detach(:device => current_attachment.device)
+        attach
+      else
+        Chef::Log.info("EBS volume #{new_resource.name} (#{aws_object.id}) is properly attached to instance #{current_attachment.instance.id} on device #{current_attachment.device}.")
+      end
+    when :available
+      attach
+    when nil
+      raise "EBS volume #{new_resource.name} does not currently exist!"
+    else
+      raise "EBS volume #{new_resource.name} (#{aws_object.id}) is in #{status} state!"
+    end
+    new_resource.save_managed_entry(aws_object, action_handler)
+  end
+
+  action :detach do
+    aws_object = new_resource.aws_object
+    status = aws_object ? aws_object.status : nil
+    case status
+    when nil
+      Chef::Log.warn "EBS volume #{new_resource.name} does not currently exist!"
+    when :in_use
+      detach
+    end
+    new_resource.save_managed_entry(aws_object, action_handler)
```

Comment:
I believe you only need this `save_managed_entry` call on the `action :create`.  The managed entry only contains AWS identifiable information so we don't need to save it on all actions.  We also don't want to create a managed entry if the user has supplied `volume_id` in the resource - because we have the unique identifier in that situation we don't need the managed entry.